src/main.rs

```diff
@@ -240,17 +240,107 @@ enum JobCancelBehaviour {
     LeaveRunning,
 }
 
+struct LocalArtifactStore {
+    lava: Arc<ThrottledLava>,
+    masker: Arc<Masker>,
+}
+
+impl LocalArtifactStore {
+    pub fn new(lava: Arc<ThrottledLava>, masker: Arc<Masker>) -> Self {
+        Self { lava, masker }
+    }
+
+    pub fn get_log(
+        &self,
+        id: i64,
+    ) -> impl Stream<Item = Result<Bytes, std::io::Error>> + Unpin + '_ {
+        self.masker.mask_stream(Box::pin(
+            self.lava
+                .log(id)
+                .map(|x| x.raw())
+                .flatten_stream()
+                .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e)),
+        ))
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+enum LavaUploadableFileType {
+    Log { id: i64 },
+}
+
+#[derive(Clone)]
+struct LavaUploadableFile {
+    store: Arc<LocalArtifactStore>,
+    which: LavaUploadableFileType,
+}
+
+impl core::fmt::Debug for LavaUploadableFile {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        self.which.fmt(f)
+    }
+}
+
+impl core::cmp::PartialEq for LavaUploadableFile {
+    fn eq(&self, other: &Self) -> bool {
+        self.which.eq(&other.which)
+    }
+}
+
+impl core::cmp::Eq for LavaUploadableFile {}
```

Comment:
Unfortunately you can't derive this, but you get errors. It appears to be an internal macro check for `Eq` (`AssertParamIsEq` is what fails for `Arc<AvailableArtifactStore>` because `AvailableArtifactStore` is not `Eq` (though it does not need to be)